Zuordnen von Datensätzen zu Herausgebern

jze opened the pull request:
Hier ist ein erster Diskussionsstand für die Implementierung der Zuordnen von Datensätzen zu Herausgebern.

Wie beim CKAN-CKAN-Harvester wird die Konfiguration `remote_orgs` verwendet, um zu steuern, ob überhaupt eine Suche nach Herausgebern stattfinden soll und wenn ja, ob auch fehlende Herausgeber automatisch angelegt werden sollen.

Als Name wir zunächst der Titel der Organsation gesucht. Gibt es keine passende Organisation, wird in Extra-Feldern `alternate_name_x` nach weiteren Bezeichnungen einer Organisation gesucht.

Fixes #19

jze committed:
consider multiple publishers in one catalog file

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -120,6 +122,52 @@ def after_update(self, harvest_object, dataset_dict, temp_dict):
+                        # At this point a configuration of the Harvest Source should be considered if there should be an error if the publisher is missing.
+                        raise ValueError("We need a new publisher with the name "+ str(publisher_name))
+
+                dataset_dict['owner_org'] = owner_org
```
Das `dataset_dict` könnte bereits einen Eintrag für `owner_org` haben, falls das in der Harvester-Konfiguration so hinterlegt wurde. Siehe [https://github.com/ckan/ckanext-dcat/blob/eb1aee013687a40d0366a2fe44546917ee960bfc/ckanext/dcat/harvesters/rdf.py#L226]. Priorität hätten in dem Fall die `remote_orgs`. In dem Falle wäre es zwar eine widersprüchliche Konfiguration, aber ich bin mir nicht sicher, ob das nicht anders herum besser wäre.

jze replied:
Beim Aufruf dieser Methode hat jeder Datensatz schon einen Eintrag in `owner_org`, nämlich die der Harvest Source.

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -26,7 +27,8 @@
 from ckanext.dcatde.validation.shacl_validation import ShaclValidator
 from ckanext.harvest.model import HarvestObject, HarvestObjectExtra
-
+from ckan.lib.search.common import SearchIndexError
```
Nicht genutzter Import, der entfernt werden kann.

jze replied:
entfernt

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -26,7 +27,8 @@
 from ckanext.harvest.model import HarvestObject, HarvestObjectExtra
-
+from ckan.lib.search.common import SearchIndexError
+from pysolr import SolrError
```
Nicht genutzter Import, der entfernt werden kann.

jze replied:
entfernt

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -120,6 +122,52 @@ def after_update(self, harvest_object, dataset_dict, temp_dict):
+        base_context = {'model': model, 'session': model.Session,
+                'user': self._get_user_name()}
+
+        remote_orgs = json.loads(harvest_object.job.source.config).get('remote_orgs', None)
```
Sollte vorher in der Methode `validate_config` [https://github.com/GovDataOfficial/ckanext-dcatde/blob/873db9a980da72859f99bb4d23c9a27c74f0424a/ckanext/dcatde/harvesters/dcatde_rdf.py#L407] geprüft werden, so dass zumindest kein ungültiges JSON abgespeichert werden kann. Ansonsten könnte es hier zu einer Exception kommen.

seitenbau-govdata commented:
Danke @jze für das Einreichen des Pull-Requests. Nach einem kurzen Durchsehen der Änderungen haben wir die ersten Anmerkungen abgegeben. Eventuell ist es später zu überlegen, ob die Erweiterung nicht sogar direkt in `ckanext-dcat` integriert werden sollte.

jze committed:
eigene Funktion zum Zuordnen des Herausgebers
überflüssige Imports entfernt
_gen_new_name Funktion verwendet
bessere Fehlermeldung bei fehlendem Herausgeber

jze committed:
determine the publisher also for updates

jze commented:
Ich habe die Stelle, an der nach Herausgebern gesucht wird, nochmal geändert. Statt in `before_create` wird die Suche nun in `_amend_package` angestoßen. So wird die Zuordnung auch beim Aktualisieren durchgeführt. Andernfalls würde beim Aktualisieren eines Datensatzes der Herausgeber wieder auf den Besitzer der Harvest-Source geändert.

In der `import_stage` prüfe ich nach dem `_amend_package`, ob die `owner_org` gesetzt wurde. Das scheint mir aber nicht besonders elegant zu sein. Wenn man an der Stelle aber nicht `False` zurückliefert, wird zwar eine Fehlermeldung im Harvestlog hinterlegt, der Datensatz würde aber auch dann angelegt werden, wenn kein Herausgeber gefunden wird.

jze committed:
Merge branch 'GovDataOfficial:master' into master

jze committed:
Merge branch 'GovDataOfficial:master' into master

jze committed:
Merge branch 'master' of https://github.com/GovDataOfficial/ckanext-dcatde

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -396,6 +445,9 @@ def import_stage(self, harvest_object):
 
         # set custom field and perform other fixes on the data
         self._amend_package(harvest_object)
+        package = json.loads(harvest_object.content)
+        if package.get('owner_org') == None:
+            return False
```
Somit würde der Datensatz auch nicht importiert werden, wenn die Konfiguration remote_orgs nicht verwendet wird. Hier könnte noch einmal geprüft werden, ob die Konfiguration gesetzt ist.

seitenbau-govdata replied:
```suggestion
        # Ensure that the property "owner_org" is set for the dataset, if the option "remote_orgs" is set, otherwise skip the import.
        if self._get_remote_orgs_from_config(harvest_object.source.config) in ('only_local', 'create'):
            package = json.loads(harvest_object.content)
            if package.get('owner_org') == None:
                return False
```
And a Getter method could be added as follows, with can be called multiple from several places.
```
    @staticmethod
    def _get_remote_orgs_from_config(source_config):
        ''' Get remote_orgs from source '''
        if source_config:
            return json.loads(source_config).get(CONFIG_PARAM_REMOTE_ORGS)

        return ''
```
The definition of the constant `CONFIG_PARAM_REMOTE_ORGS` used in the example is mandatory.

seitenbau-govdata commented on ckanext/dcatde/harvesters/dcatde_rdf.py:
```diff
@@ -114,6 +114,53 @@ def after_parsing(self, rdf_parser, harvest_job):
+                           owner_org = org.group_id
+
+                if not owner_org:
+                    if remote_orgs == 'create':
```
Theoretisch könnte es vorkommen, dass "fälschlicherweise" mehrere Organisationen mit demselben Titel existieren. Bei `len(organizations) > 1` , wird dann eine neue Organisation angelegt. In diesem Fall wäre es vielleicht sinnvoll, eine Fehlermeldung im Harvestlog zu hinterlegen oder stattdessen die erste ID aus der `organizations`-Liste zuzuweisen.

thorge committed:
Merge branch 'GovDataOfficial:master' into master

thorge committed:
Refactor organization slug generation and remote_orgs handling in DCATdeRDFHarvester

- Introduced `_slugify` static method to normalize publisher names, handling umlauts
  and other non-ASCII characters for CKAN organization slugs.
- Replaced legacy `_gen_new_name` logic with `_slugify` when creating new organizations.
- Added `_get_remote_orgs_from_config` static method to safely read `remote_orgs` from source config.
- Updated `_assign_owner_org` to use `_slugify` and `_get_remote_orgs_from_config`.

jze committed:
Merge branch 'GovDataOfficial:master' into master
